File: packages/merkle-tree/src/test/iterators.test.ts
import { 
    iterateNodes, 
    iterateLeaves,
    SortNode,
} from '../lib/merkle-tree';
import { buildTree, leaf, node } from './merkle-verify';

describe('iterateNodes', () => {
    
    describe('Edge Cases', () => {
        test('returns nothing for undefined node', () => {
            const nodes = Array.from(iterateNodes(undefined));
            expect(nodes).toEqual([]);
        });

        test('iterates single leaf node', () => {
            const sortTree = leaf('A');
            const nodes = Array.from(iterateNodes<SortNode>(sortTree));
            
            expect(nodes.length).toBe(1);
            expect(nodes[0].fileName).toBe('A');
            expect(nodes[0].contentHash).toEqual(Buffer.from('A'));
        });

        test('iterates two leaf nodes', () => {
            const sortTree = node(leaf('A'), leaf('B'));
            const nodes = Array.from(iterateNodes<SortNode>(sortTree));
            
            // Should return parent node, then left child, then right child
            expect(nodes.length).toBe(3);
            expect(nodes[0].minFileName).toBe('A'); // Parent node
            expect(nodes[1].fileName).toBe('A'); // Left leaf
            expect(nodes[2].fileName).toBe('B'); // Right leaf
        });
    });

    describe('Traversal Order', () => {
        test('performs pre-order traversal (parent, left, right)', () => {
            // Build tree: ((A, B), C)
            const sortTree = node(node(leaf('A'), leaf('B')), leaf('C'));
            const nodes = Array.from(iterateNodes<SortNode>(sortTree));
            
            // Pre-order: root, left subtree (parent, A, B), right (C)
            expect(nodes.length).toBe(5);
            
            // First node is root
            expect(nodes[0].minFileName).toBe('A');
            expect(nodes[0].nodeCount).toBe(5);
            
            // Second node is left child (parent of A and B)
            expect(nodes[1].minFileName).toBe('A');
            expect(nodes[1].nodeCount).toBe(3);
            
            // Third node is leaf A
            expect(nodes[2].fileName).toBe('A');
            
            // Fourth node is leaf B
            expect(nodes[3].fileName).toBe('B');
            
            // Fifth node is leaf C
            expect(nodes[4].fileName).toBe('C');
        });

        test('visits all nodes in correct order for balanced tree', () => {
            // Build tree with 4 leaves: ((A, B), (C, D))
            const tree = buildTree(['A', 'B', 'C', 'D']);
            const nodes = Array.from(iterateNodes<SortNode>(tree.sort!));
            
            // Should have 7 nodes total: 1 root + 2 internal + 4 leaves
            expect(nodes.length).toBe(7);
            
            // Verify all nodes are present
            const leafNodes = nodes.filter(n => n.fileName !== undefined);
            expect(leafNodes.length).toBe(4);
            expect(leafNodes.map(n => n.fileName)).toEqual(['A', 'B', 'C', 'D']);
        });
    });

    describe('Node Count', () => {
        test('counts all nodes correctly for small tree', () => {
            const tree = buildTree(['A', 'B', 'C']);
            const nodes = Array.from(iterateNodes<SortNode>(tree.sort!));
            
            // Tree structure: ((A, B), C) = 5 nodes total
            expect(nodes.length).toBe(5);
        });

        test('counts all nodes correctly for power of 2 leaves', () => {
            const tree = buildTree(['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']);
            const nodes = Array.from(iterateNodes<SortNode>(tree.sort!));
            
            // For 8 leaves in a binary tree, we have 15 total nodes (8 leaves + 7 internal)
            expect(nodes.length).toBe(15);
        });

        test('counts all nodes correctly for non-power of 2 leaves', () => {
            const tree = buildTree(['A', 'B', 'C', 'D', 'E']);
            const nodes = Array.from(iterateNodes<SortNode>(tree.sort!));
            
            // For 5 leaves, we should have 9 total nodes
            expect(nodes.length).toBe(9);
        });
    });

    describe('Node Properties', () => {
        test('all nodes have valid nodeCount property', () => {
            const tree = buildTree(['A', 'B', 'C', 'D', 'E', 'F']);
            const nodes = Array.from(iterateNodes<SortNode>(tree.sort!));
            
            for (const node of nodes) {
                expect(node.nodeCount).toBeGreaterThanOrEqual(1);
                expect(typeof node.nodeCount).toBe('number');
            }
        });

        test('all nodes have valid minFileName property', () => {
            const tree = buildTree(['A', 'B', 'C', 'D']);
            const nodes = Array.from(iterateNodes<SortNode>(tree.sort!));
            
            for (const node of nodes) {
                expect(node.minFileName).toBeDefined();
                expect(typeof node.minFileName).toBe('string');
            }
        });

        test('leaf nodes have fileName property', () => {
            const tree = buildTree(['A', 'B', 'C']);
            const nodes = Array.from(iterateNodes<SortNode>(tree.sort!));
            
            const leafNodes = nodes.filter(n => n.fileName !== undefined);
            expect(leafNodes.length).toBe(3);
            
            for (const leaf of leafNodes) {
                expect(leaf.contentHash).toBeDefined();
                expect(leaf.nodeCount).toBe(1);
            }
        });
    });

    describe('Large Trees', () => {
        test('iterates large tree with 100 nodes', () => {
            const fileNames = Array.from({ length: 100 }, (_, i) => 
                `file_${i.toString().padStart(3, '0')}`
            );
            const tree = buildTree(fileNames);
            const nodes = Array.from(iterateNodes<SortNode>(tree.sort!));
            
            // Should have all nodes
            expect(nodes.length).toBeGreaterThanOrEqual(100);
            
            // All 100 files should be present
            const leafNodes = nodes.filter(n => n.fileName !== undefined);
            expect(leafNodes.length).toBe(100);
        });
    });

    describe('Generator Behavior', () => {
        test('can be used in for...of loop', () => {
            const tree = buildTree(['A', 'B', 'C']);
            const nodeNames: string[] = [];
            
            for (const node of iterateNodes<SortNode>(tree.sort!)) {
                nodeNames.push(node.minFileName);
            }
            
            expect(nodeNames.length).toBeGreaterThan(0);
        });

        test('can be converted to array multiple times', () => {
            const tree = buildTree(['A', 'B', 'C']);
            
            const nodes1 = Array.from(iterateNodes<SortNode>(tree.sort!));
            const nodes2 = Array.from(iterateNodes<SortNode>(tree.sort!));
            
            expect(nodes1.length).toBe(nodes2.length);
        });

        test('is lazy and does not iterate until consumed', () => {
            const tree = buildTree(['A', 'B', 'C']);
            
            // Creating the generator should not iterate
            const generator = iterateNodes<SortNode>(tree.sort!);
            
            // Only when we consume it does it iterate
            const firstNode = generator.next();
            expect(firstNode.done).toBe(false);
            expect(firstNode.value).toBeDefined();
        });
    });
});

describe('iterateLeaves', () => {
    
    describe('Edge Cases', () => {
        test('returns nothing for undefined node', () => {
            const leaves = Array.from(iterateLeaves(undefined));
            expect(leaves).toEqual([]);
        });

        test('iterates single leaf node', () => {
            const sortTree = leaf('A');
            const leaves = Array.from(iterateLeaves<SortNode>(sortTree));
            
            expect(leaves.length).toBe(1);
            expect(leaves[0].fileName).toBe('A');
            expect(leaves[0].contentHash).toEqual(Buffer.from('A'));
        });

        test('iterates two leaf nodes', () => {
            const sortTree = node(leaf('A'), leaf('B'));
            const leaves = Array.from(iterateLeaves<SortNode>(sortTree));
            
            expect(leaves.length).toBe(2);
            expect(leaves[0].fileName).toBe('A');
            expect(leaves[1].fileName).toBe('B');
        });
    });

    describe('Leaf-Only Filtering', () => {
        test('only returns leaf nodes, not internal nodes', () => {
            const tree = buildTree(['A', 'B', 'C', 'D']);
            const leaves = Array.from(iterateLeaves<SortNode>(tree.sort!));
            
            // Should only have leaf nodes
            expect(leaves.length).toBe(4);
            
            // All should have fileName property (leaf indicator)
            for (const leaf of leaves) {
                expect(leaf.fileName).toBeDefined();
                expect(leaf.contentHash).toBeDefined();
                expect(leaf.nodeCount).toBe(1);
            }
        });

        test('does not return parent nodes', () => {
            const tree = buildTree(['A', 'B', 'C', 'D', 'E']);
            const leaves = Array.from(iterateLeaves<SortNode>(tree.sort!));
            const allNodes = Array.from(iterateNodes<SortNode>(tree.sort!));
            
            // Leaves should be fewer than all nodes
            expect(leaves.length).toBeLessThan(allNodes.length);
            expect(leaves.length).toBe(5);
        });
    });

    describe('Leaf Order Preservation', () => {
        test('preserves in-order traversal of leaves', () => {
            const tree = buildTree(['A', 'B', 'C', 'D']);
            const leaves = Array.from(iterateLeaves<SortNode>(tree.sort!));
            
            const fileNames = leaves.map(l => l.fileName);
            expect(fileNames).toEqual(['A', 'B', 'C', 'D']);
        });

        test('maintains sorted order for unsorted input', () => {
            const tree = buildTree(['D', 'A', 'C', 'B', 'E']);
            const leaves = Array.from(iterateLeaves<SortNode>(tree.sort!));
            
            const fileNames = leaves.map(l => l.fileName);
            // Should be sorted because buildTree creates a sorted tree
            expect(fileNames).toEqual(['A', 'B', 'C', 'D', 'E']);
        });

        test('maintains order with UUID filenames', () => {
            const fileNames = [
                'asset/3e4f1677-dfc1-4efe-be57-6969e0b1c9b6',
                'asset/7b4f6865-26a5-4316-98ba-41e528594ec0',
                'asset/7c86cb29-c6ee-40dc-9d08-a8dc5c5a0dc7',
            ];
            
            const tree = buildTree(fileNames);
            const leaves = Array.from(iterateLeaves<SortNode>(tree.sort!));
            
            expect(leaves.map(l => l.fileName)).toEqual(fileNames);
        });
    });

    describe('Leaf Count', () => {
        test('counts leaves correctly for power of 2', () => {
            const tree = buildTree(['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']);
            const leaves = Array.from(iterateLeaves<SortNode>(tree.sort!));
            
            expect(leaves.length).toBe(8);
        });

        test('counts leaves correctly for non-power of 2', () => {
            const tree = buildTree(['A', 'B', 'C', 'D', 'E', 'F', 'G']);
            const leaves = Array.from(iterateLeaves<SortNode>(tree.sort!));
            
            expect(leaves.length).toBe(7);
        });

        test('leaf count matches tree leafCount property', () => {
            const tree = buildTree(['A', 'B', 'C', 'D', 'E', 'F']);
            const leaves = Array.from(iterateLeaves<SortNode>(tree.sort!));
            
            expect(leaves.length).toBe(tree.sort!.leafCount);
        });
    });

    describe('Leaf Properties', () => {
        test('all leaves have contentHash', () => {
            const tree = buildTree(['A', 'B', 'C', 'D']);
            const leaves = Array.from(iterateLeaves<SortNode>(tree.sort!));
            
            for (const leaf of leaves) {
                expect(leaf.contentHash).toBeDefined();
                expect(Buffer.isBuffer(leaf.contentHash)).toBe(true);
            }
        });

        test('all leaves have fileName', () => {
            const tree = buildTree(['A', 'B', 'C', 'D']);
            const leaves = Array.from(iterateLeaves<SortNode>(tree.sort!));
            
            for (const leaf of leaves) {
                expect(leaf.fileName).toBeDefined();
                expect(typeof leaf.fileName).toBe('string');
            }
        });

        test('all leaves have nodeCount of 1', () => {
            const tree = buildTree(['A', 'B', 'C', 'D', 'E']);
            const leaves = Array.from(iterateLeaves<SortNode>(tree.sort!));
            
            for (const leaf of leaves) {
                expect(leaf.nodeCount).toBe(1);
            }
        });

        test('all leaves have leafCount of 1', () => {
            const tree = buildTree(['A', 'B', 'C', 'D', 'E']);
            const leaves = Array.from(iterateLeaves<SortNode>(tree.sort!));
            
            for (const leaf of leaves) {
                expect(leaf.leafCount).toBe(1);
            }
        });

        test('all leaves have no children', () => {
            const tree = buildTree(['A', 'B', 'C', 'D']);
            const leaves = Array.from(iterateLeaves<SortNode>(tree.sort!));
            
            for (const leaf of leaves) {
                expect(leaf.left).toBeUndefined();
                expect(leaf.right).toBeUndefined();
            }
        });
    });

    describe('Large Trees', () => {
        test('iterates 100 leaves efficiently', () => {
            const fileNames = Array.from({ length: 100 }, (_, i) => 
                `file_${i.toString().padStart(3, '0')}`
            );
            const tree = buildTree(fileNames);
            const leaves = Array.from(iterateLeaves<SortNode>(tree.sort!));
            
            expect(leaves.length).toBe(100);
            
            // Verify all files are present
            const leafFileNames = leaves.map(l => l.fileName);
            expect(leafFileNames).toEqual(fileNames);
        });

        test('iterates 1000 leaves efficiently', () => {
            const fileNames = Array.from({ length: 1000 }, (_, i) => 
                `file_${i.toString().padStart(4, '0')}`
            );
            const tree = buildTree(fileNames);
            const leaves = Array.from(iterateLeaves<SortNode>(tree.sort!));
            
            expect(leaves.length).toBe(1000);
        });
    });

    describe('Generator Behavior', () => {
        test('can be used in for...of loop', () => {
            const tree = buildTree(['A', 'B', 'C']);
            const fileNames: string[] = [];
            
            for (const leaf of iterateLeaves<SortNode>(tree.sort!)) {
                fileNames.push(leaf.fileName!);
            }
            
            expect(fileNames).toEqual(['A', 'B', 'C']);
        });

        test('can be converted to array multiple times', () => {
            const tree = buildTree(['A', 'B', 'C']);
            
            const leaves1 = Array.from(iterateLeaves<SortNode>(tree.sort!));
            const leaves2 = Array.from(iterateLeaves<SortNode>(tree.sort!));
            
            expect(leaves1.length).toBe(leaves2.length);
            expect(leaves1.map(l => l.fileName)).toEqual(leaves2.map(l => l.fileName));
        });

        test('is lazy and does not iterate until consumed', () => {
            const tree = buildTree(['A', 'B', 'C']);
            
            // Creating the generator should not iterate
            const generator = iterateLeaves<SortNode>(tree.sort!);
            
            // Only when we consume it does it iterate
            const firstLeaf = generator.next();
            expect(firstLeaf.done).toBe(false);
            expect(firstLeaf.value).toBeDefined();
            expect(firstLeaf.value.fileName).toBe('A');
        });

        test('can be partially consumed', () => {
            const tree = buildTree(['A', 'B', 'C', 'D', 'E']);
            const generator = iterateLeaves<SortNode>(tree.sort!);
            
            // Consume first 3 leaves
            const leaf1 = generator.next();
            const leaf2 = generator.next();
            const leaf3 = generator.next();
            
            expect(leaf1.value.fileName).toBe('A');
            expect(leaf2.value.fileName).toBe('B');
            expect(leaf3.value.fileName).toBe('C');
            
            // Can still get remaining leaves
            const leaf4 = generator.next();
            const leaf5 = generator.next();
            const done = generator.next();
            
            expect(leaf4.value.fileName).toBe('D');
            expect(leaf5.value.fileName).toBe('E');
            expect(done.done).toBe(true);
        });
    });

    describe('Comparison with Manual Collection', () => {
        test('produces same results as manual recursive collection', () => {
            const tree = buildTree(['A', 'B', 'C', 'D', 'E', 'F']);
            
            // Manual collection using recursion (like in the existing tests)
            const manualLeaves: SortNode[] = [];
            function collectLeaves(node: SortNode | undefined) {
                if (!node) return;
                if (!node.left && !node.right) {
                    manualLeaves.push(node);
                } else {
                    collectLeaves(node.left);
                    collectLeaves(node.right);
                }
            }
            collectLeaves(tree.sort);
            
            // Using iterateLeaves
            const iteratorLeaves = Array.from(iterateLeaves<SortNode>(tree.sort!));
            
            // Both should produce the same results
            expect(iteratorLeaves.length).toBe(manualLeaves.length);
            expect(iteratorLeaves.map(l => l.fileName)).toEqual(manualLeaves.map(l => l.fileName));
        });
    });

    describe('Empty Subtrees', () => {
        test('handles nodes with only left child', () => {
            // Manually construct a tree with imbalanced structure
            const leftLeaf = leaf('A');
            const parent: SortNode = {
                nodeCount: 2,
                leafCount: 1,
                size: leftLeaf.size,
                minFileName: 'A',
                left: leftLeaf,
                // right is undefined
            };
            
            const leaves = Array.from(iterateLeaves<SortNode>(parent));
            expect(leaves.length).toBe(1);
            expect(leaves[0].fileName).toBe('A');
        });

        test('handles nodes with only right child', () => {
            // Manually construct a tree with only right child
            const rightLeaf = leaf('B');
            const parent: SortNode = {
                nodeCount: 2,
                leafCount: 1,
                size: rightLeaf.size,
                minFileName: 'B',
                right: rightLeaf,
                // left is undefined
            };
            
            const leaves = Array.from(iterateLeaves<SortNode>(parent));
            expect(leaves.length).toBe(1);
            expect(leaves[0].fileName).toBe('B');
        });
    });
});

describe('iterateNodes vs iterateLeaves', () => {
    test('iterateNodes returns more nodes than iterateLeaves', () => {
        const tree = buildTree(['A', 'B', 'C', 'D', 'E']);
        
        const allNodes = Array.from(iterateNodes<SortNode>(tree.sort!));
        const leaves = Array.from(iterateLeaves<SortNode>(tree.sort!));
        
        expect(allNodes.length).toBeGreaterThan(leaves.length);
    });

    test('all leaves from iterateLeaves are present in iterateNodes', () => {
        const tree = buildTree(['A', 'B', 'C', 'D']);
        
        const allNodes = Array.from(iterateNodes<SortNode>(tree.sort!));
        const leaves = Array.from(iterateLeaves<SortNode>(tree.sort!));
        
        const leafFileNames = leaves.map(l => l.fileName);
        const nodesWithFiles = allNodes.filter(n => n.fileName !== undefined);
        const nodeFileNames = nodesWithFiles.map(n => n.fileName);
        
        expect(leafFileNames).toEqual(nodeFileNames);
    });

    test('difference in count equals number of internal nodes', () => {
        const tree = buildTree(['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']);
        
        const allNodes = Array.from(iterateNodes<SortNode>(tree.sort!));
        const leaves = Array.from(iterateLeaves<SortNode>(tree.sort!));
        
        const internalNodeCount = allNodes.length - leaves.length;
        
        // For a binary tree with n leaves, we have n-1 internal nodes
        // (but this can vary based on tree structure)
        expect(internalNodeCount).toBeGreaterThan(0);
        expect(internalNodeCount).toBeLessThan(leaves.length);
    });
});

File: packages/merkle-tree-zig/src/test/iterators.test.zig
//
// Tests for iterateLeaves (port of src/test/iterators.test.ts; the iterateNodes tests are not ported
// because iterateNodes is not ported).
// (Zig: test names are prefixed with the describe name.)
//

const std = @import("std");
const merkle_tree_zig = @import("merkle-tree-zig");
const merkle_verify = @import("merkle-verify.zig");
const merkle_tree = merkle_tree_zig.merkle_tree;
const SortNode = merkle_tree.SortNode;
const buildTree = merkle_verify.buildTree;
const leaf = merkle_verify.leaf;
const node = merkle_verify.node;

//
// Collects every node iterateLeaves yields.
//
fn allLeaves(allocator: std.mem.Allocator, root: ?*SortNode) ![]const *SortNode {
    var nodes: std.ArrayList(*SortNode) = .empty;
    var iterator = merkle_tree.iterateLeaves(SortNode, allocator, root);
    while (try iterator.next()) |current| {
        try nodes.append(allocator, current);
    }
    return nodes.items;
}

//
// The names of the nodes that have a name.
//
fn namesOf(allocator: std.mem.Allocator, nodes: []const *SortNode) ![]const []const u8 {
    var names: std.ArrayList([]const u8) = .empty;
    for (nodes) |current| {
        if (current.name) |name| {
            try names.append(allocator, name);
        }
    }
    return names.items;
}

//
// Checks two lists of names are equal.
//
fn expectNames(expected: []const []const u8, actual: []const []const u8) !void {
    try std.testing.expectEqual(expected.len, actual.len);
    for (expected, actual) |expectedName, actualName| {
        try std.testing.expectEqualStrings(expectedName, actualName);
    }
}

//
// Names like `file_000` (TypeScript: `file_${i.toString().padStart(3, '0')}`).
//
fn paddedFileNames(allocator: std.mem.Allocator, count: usize, width: usize) ![]const []const u8 {
    const names = try allocator.alloc([]const u8, count);
    for (names, 0..) |*name, index| {
        const digits = try std.fmt.allocPrint(allocator, "{d}", .{index});
        const padding = try allocator.alloc(u8, width -| digits.len);
        @memset(padding, '0');
        name.* = try std.fmt.allocPrint(allocator, "file_{s}{s}", .{ padding, digits });
    }
    return names;
}

test "iterateLeaves: returns nothing for undefined node" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    try std.testing.expectEqual(@as(usize, 0), (try allLeaves(arena.allocator(), null)).len);
}

test "iterateLeaves: iterates single leaf node" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const leaves = try allLeaves(allocator, try leaf(allocator, "A", 100));
    try std.testing.expectEqual(@as(usize, 1), leaves.len);
    try std.testing.expectEqualStrings("A", leaves[0].name.?);
    try std.testing.expectEqualStrings("A", leaves[0].contentHash.?);
}

test "iterateLeaves: iterates two leaf nodes" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const leaves = try allLeaves(allocator, try node(allocator, try leaf(allocator, "A", 100), try leaf(allocator, "B", 100)));
    try expectNames(&.{ "A", "B" }, try namesOf(allocator, leaves));
}

test "iterateLeaves: only returns leaf nodes, not internal nodes" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const tree = try buildTree(allocator, &.{ "A", "B", "C", "D" });
    const leaves = try allLeaves(allocator, tree.sort);
    try std.testing.expectEqual(@as(usize, 4), leaves.len);
    for (leaves) |current| {
        try std.testing.expect(current.name != null);
        try std.testing.expect(current.contentHash != null);
        try std.testing.expectEqual(@as(u32, 1), current.nodeCount);
    }
}

test "iterateLeaves: does not return parent nodes" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const tree = try buildTree(allocator, &.{ "A", "B", "C", "D", "E" });
    const leaves = try allLeaves(allocator, tree.sort);
    try std.testing.expect(leaves.len < tree.sort.?.nodeCount);
    try std.testing.expectEqual(@as(usize, 5), leaves.len);
}

test "iterateLeaves: preserves in-order traversal of leaves" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const tree = try buildTree(allocator, &.{ "A", "B", "C", "D" });
    try expectNames(&.{ "A", "B", "C", "D" }, try namesOf(allocator, try allLeaves(allocator, tree.sort)));
}

test "iterateLeaves: maintains sorted order for unsorted input" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const tree = try buildTree(allocator, &.{ "D", "A", "C", "B", "E" });
    try expectNames(&.{ "A", "B", "C", "D", "E" }, try namesOf(allocator, try allLeaves(allocator, tree.sort)));
}

test "iterateLeaves: maintains order with UUID filenames" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const fileNames = [_][]const u8{
        "asset/3e4f1677-dfc1-4efe-be57-6969e0b1c9b6",
        "asset/7b4f6865-26a5-4316-98ba-41e528594ec0",
        "asset/7c86cb29-c6ee-40dc-9d08-a8dc5c5a0dc7",
    };
    const tree = try buildTree(allocator, &fileNames);
    try expectNames(&fileNames, try namesOf(allocator, try allLeaves(allocator, tree.sort)));
}

test "iterateLeaves: counts leaves correctly for power of 2" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const tree = try buildTree(allocator, &.{ "A", "B", "C", "D", "E", "F", "G", "H" });
    try std.testing.expectEqual(@as(usize, 8), (try allLeaves(allocator, tree.sort)).len);
}

test "iterateLeaves: counts leaves correctly for non-power of 2" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const tree = try buildTree(allocator, &.{ "A", "B", "C", "D", "E", "F", "G" });
    try std.testing.expectEqual(@as(usize, 7), (try allLeaves(allocator, tree.sort)).len);
}

test "iterateLeaves: leaf count matches tree leafCount property" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const tree = try buildTree(allocator, &.{ "A", "B", "C", "D", "E", "F" });
    try std.testing.expectEqual(@as(usize, tree.sort.?.leafCount), (try allLeaves(allocator, tree.sort)).len);
}

test "iterateLeaves: all leaves have contentHash" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const tree = try buildTree(allocator, &.{ "A", "B", "C", "D" });
    for (try allLeaves(allocator, tree.sort)) |current| {
        try std.testing.expect(current.contentHash != null);
    }
}

test "iterateLeaves: all leaves have name" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const tree = try buildTree(allocator, &.{ "A", "B", "C", "D" });
    for (try allLeaves(allocator, tree.sort)) |current| {
        try std.testing.expect(current.name != null);
    }
}

test "iterateLeaves: all leaves have nodeCount of 1" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const tree = try buildTree(allocator, &.{ "A", "B", "C", "D", "E" });
    for (try allLeaves(allocator, tree.sort)) |current| {
        try std.testing.expectEqual(@as(u32, 1), current.nodeCount);
    }
}

test "iterateLeaves: all leaves have leafCount of 1" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const tree = try buildTree(allocator, &.{ "A", "B", "C", "D", "E" });
    for (try allLeaves(allocator, tree.sort)) |current| {
        try std.testing.expectEqual(@as(u32, 1), current.leafCount);
    }
}

test "iterateLeaves: all leaves have no children" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const tree = try buildTree(allocator, &.{ "A", "B", "C", "D" });
    for (try allLeaves(allocator, tree.sort)) |current| {
        try std.testing.expect(current.left == null);
        try std.testing.expect(current.right == null);
    }
}

test "iterateLeaves: iterates 100 leaves efficiently" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const fileNames = try paddedFileNames(allocator, 100, 3);
    const tree = try buildTree(allocator, fileNames);
    try expectNames(fileNames, try namesOf(allocator, try allLeaves(allocator, tree.sort)));
}

test "iterateLeaves: iterates 1000 leaves efficiently" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const tree = try buildTree(allocator, try paddedFileNames(allocator, 1000, 4));
    try std.testing.expectEqual(@as(usize, 1000), (try allLeaves(allocator, tree.sort)).len);
}

test "iterateLeaves: can be used in for...of loop" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const tree = try buildTree(allocator, &.{ "A", "B", "C" });
    try expectNames(&.{ "A", "B", "C" }, try namesOf(allocator, try allLeaves(allocator, tree.sort)));
}

test "iterateLeaves: can be converted to array multiple times" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const tree = try buildTree(allocator, &.{ "A", "B", "C" });
    const leaves1 = try namesOf(allocator, try allLeaves(allocator, tree.sort));
    const leaves2 = try namesOf(allocator, try allLeaves(allocator, tree.sort));
    try expectNames(leaves1, leaves2);
}

test "iterateLeaves: is lazy and does not iterate until consumed" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const tree = try buildTree(allocator, &.{ "A", "B", "C" });
    var iterator = merkle_tree.iterateLeaves(SortNode, allocator, tree.sort);
    try std.testing.expectEqual(@as(usize, 0), iterator.stack.items.len);
    const firstLeaf = try iterator.next();
    try std.testing.expectEqualStrings("A", firstLeaf.?.name.?);
}

test "iterateLeaves: can be partially consumed" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const tree = try buildTree(allocator, &.{ "A", "B", "C", "D", "E" });
    var iterator = merkle_tree.iterateLeaves(SortNode, allocator, tree.sort);

    // Consume first 3 leaves
    try std.testing.expectEqualStrings("A", (try iterator.next()).?.name.?);
    try std.testing.expectEqualStrings("B", (try iterator.next()).?.name.?);
    try std.testing.expectEqualStrings("C", (try iterator.next()).?.name.?);

    // Can still get remaining leaves
    try std.testing.expectEqualStrings("D", (try iterator.next()).?.name.?);
    try std.testing.expectEqualStrings("E", (try iterator.next()).?.name.?);
    try std.testing.expect((try iterator.next()) == null);
}

//
// Collects leaves by manual recursion (like the existing tests).
//
fn collectLeaves(allocator: std.mem.Allocator, leaves: *std.ArrayList(*SortNode), sortNode: ?*SortNode) !void {
    const currentNode = sortNode orelse {
        return;
    };
    if (currentNode.left == null and currentNode.right == null) {
        try leaves.append(allocator, currentNode);
    }
    else {
        try collectLeaves(allocator, leaves, currentNode.left);
        try collectLeaves(allocator, leaves, currentNode.right);
    }
}

test "iterateLeaves: produces same results as manual recursive collection" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const tree = try buildTree(allocator, &.{ "A", "B", "C", "D", "E", "F" });

    var manualLeaves: std.ArrayList(*SortNode) = .empty;
    try collectLeaves(allocator, &manualLeaves, tree.sort);
    const iteratorLeaves = try allLeaves(allocator, tree.sort);
    try expectNames(try namesOf(allocator, manualLeaves.items), try namesOf(allocator, iteratorLeaves));
}

test "iterateLeaves: handles nodes with only left child" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const leftLeaf = try leaf(allocator, "A", 100);
    var parent: SortNode = .{ .nodeCount = 2, .leafCount = 1, .size = leftLeaf.size, .minName = "A", .left = leftLeaf };
    try expectNames(&.{"A"}, try namesOf(allocator, try allLeaves(allocator, &parent)));
}

test "iterateLeaves: handles nodes with only right child" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const rightLeaf = try leaf(allocator, "B", 100);
    var parent: SortNode = .{ .nodeCount = 2, .leafCount = 1, .size = rightLeaf.size, .minName = "B", .right = rightLeaf };
    try expectNames(&.{"B"}, try namesOf(allocator, try allLeaves(allocator, &parent)));
}
